Proposal fixes

dknopik opened the pull request:
This PR fixes proposals! The commits fix several issues, I am not sure which are strictly necessary to fix proposals, but each is an improvement anyway.

[Decode `DataSsz` variants explicitly based on duty and fork](https://github.com/sigp/anchor/pull/310/commits/cf573b6e72744a8bc0bcda4f81347af43d1d7fc9): Currently, Fulu and Electra blocks are identical when serialized in SSZ. Therefore, when we deserialize/deserialise the block, `from_ssz_bytes` accidentally decodes the block into the `Fulu` variant, causing us to send a Fulu block to the BN, which in turn refuses to publish it. The commit explicitly decodes the block using the `version` field from the decided data and similarly decodes other objects directly as well in other duties (to be safe).

[Set `max_transmit_size`](https://github.com/sigp/anchor/pull/310/commits/a353880036d9baee590250b7775d8cda2373b710): Some messages with large blocks failed to be sent as they are quite large - we set `max_transmit_size` to a large enough value of 5 million bytes. 

**Outdated, see "Improve fix for round timeouts" below** - [Avoid conflict between timer-based and message-based round increments](https://github.com/sigp/anchor/pull/310/commits/159126ed89ff1097c29df2a4da674399377743cd): If we receive a quorum of round change messages, we set our own round to the round mentioned in these messages. However, if a timer tick arrives after, it increased the round regardless - even though actually it meant to increase the round to the round found in those messages. We fix this by tracking the round requested by ticks separately.

[Remember PREPARE even if we have not received the PROPOSE yet](https://github.com/sigp/anchor/pull/310/commits/f6d89986384f0d0c13683fb1708b63dbd79a6f4d): We discarded PREPARE messages if we have not received a PROPOSE message yet - which in some cases resulted in missed quorums. We record unknown prepare messages, but still only check for quorum if we have the PROPOSE message.

[Use current slot for RANDAO Reveal](https://github.com/sigp/anchor/pull/310/commits/460a288ce3f3359a3eb3c8d8d9e8131d80c65634): We need to use the current slot instead of the last slot of the epoch for go-ssv to pick up the messages correctly.

[Address review and fix proposal data](https://github.com/sigp/anchor/pull/310/commits/a309453dee9a04373e117778b7a407e4669b7661): While the logic was correct for blinded block and pre-Deneb full blocks, it was not correct for Deneb and Electra blocks as blobs need to decided along with the block. This commit sets the data appropriately - but this requires a small change in LH which is pending approval: https://github.com/sigp/lighthouse/pull/7497.

[Improve fix for round timeouts](https://github.com/sigp/anchor/pull/310/commits/73c5ce77a2b228ba841f533bc2aa223421401ae7): Round changes might be triggered by messages instead of timer ticks. In order to avoid incorrect timeouts after a round has been changed by a message (e.g. after replaying buffered messages), we calculate the currently adequate timeout every time before we wait for a message or round timeout.

dknopik committed:
Set `max_transmit_size`

dknopik committed:
Decode `DataSsz` variants explicitly based on duty and fork

dknopik committed:
Avoid conflict between timer-based and message-based round increments

dknopik committed:
Remember PREPARE even if we have not received the PROPOSE yet

dknopik committed:
fmt

diegomrsantos commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -255,6 +254,22 @@ impl TreeHash for DataVersion {
+impl TryFrom<DataVersion> for ForkName {
+    type Error = ();
+
+    fn try_from(value: DataVersion) -> Result<ForkName, Self::Error> {
```
is this related to https://github.com/sigp/anchor/pull/310/commits/cf573b6e72744a8bc0bcda4f81347af43d1d7fc9#diff-25a340f5184292e9132e4b012313bfd9e056682eb85555b96830b1909801b2f1R441?

dknopik replied:
yes: `ForkName::try_from(completed_data.version)`

diegomrsantos committed:
improve readability

dknopik commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -603,6 +601,12 @@ where
             warn!(from = ?operator_id, "PREPARE message is a duplicate")
         }
 
+        // Make sure that we have accepted a proposal for this round
+        if !self.proposal_accepted_for_current_round {
+            warn!(from=?operator_id, ?self.state, self=?self.config.operator_id(), "Have not accepted Proposal for current round yet");
+            return;
+        }
+
```
what? no. i removed the other one

dknopik commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -255,6 +254,22 @@ impl TreeHash for DataVersion {
     }
 }
 
```
This is self-explanatory. a 12 line comment explaining this very simple functionality is noise.

diegomrsantos replied:
Unfortunately, it got quite verbose with this kind of suggestion. At least, it says [nitpick].

dknopik committed:
log block decoding error

dknopik commented on anchor/validator_store/src/lib.rs:
```diff
@@ -459,10 +460,21 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+            Ok(DataSsz::BeaconBlock(block))
+        } else {
+            error!(%fork, "Failed to deserialize decided block");
+            Err(Error::SpecificError(SpecificError::InvalidQbftData))
```
fair enough, but I think one error is enough information here

diegomrsantos commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -225,7 +224,7 @@ impl TreeHash for BeaconRole {
-#[derive(Clone, Debug, PartialEq, Decode, Encode)]
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord, Decode, Encode)]
 #[ssz(struct_behaviour = "transparent")]
 pub struct DataVersion(u64);
```
Could this be an enum?

dknopik replied:
yes, and it probably should be. do you want me to change it in this PR or in a different one?

diegomrsantos replied:
I don't have any preference, whatever works better for you

dknopik replied:
let's do it separately, I am eager to refactor some types anyway :)

diegomrsantos commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -255,6 +254,22 @@ impl TreeHash for DataVersion {
+            DATA_VERSION_CAPELLA => Ok(ForkName::Capella),
+            DATA_VERSION_DENEB => Ok(ForkName::Deneb),
+            DATA_VERSION_ELECTRA => Ok(ForkName::Electra),
+            _ => Err(()),
```
How about a more descriptive error?

dknopik replied:
there is only one error case. I am usually against returning strings in `TryFrom` implementations, as the caller might not even need an error message as it can fall back to some other logic. A single variant enum is also noise IMO.

diegomrsantos replied:
how about 

```rust
#[derive(Debug)]
pub struct UnknownDataVersion(pub DataVersion);

impl TryFrom<DataVersion> for ForkName {
    type Error = UnknownDataVersion;

    fn try_from(v: DataVersion) -> Result<Self, Self::Error> {
        match v {
            DATA_VERSION_PHASE0   => Ok(ForkName::Base),
            DATA_VERSION_ALTAIR   => Ok(ForkName::Altair),
            DATA_VERSION_BELLATRIX=> Ok(ForkName::Bellatrix),
            DATA_VERSION_CAPELLA  => Ok(ForkName::Capella),
            DATA_VERSION_DENEB    => Ok(ForkName::Deneb),
            DATA_VERSION_ELECTRA  => Ok(ForkName::Electra),
            _ => Err(UnknownDataVersion(v)),
        }
    }
}
```

dknopik replied:
@diegomrsantos I've reworked the DataVersion a bit, using ForkName internally. Looking forwards to your thoughts about that approach, I am a bit on the fence about it.

diegomrsantos commented on anchor/validator_store/src/lib.rs:
```diff
@@ -413,43 +414,55 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+            .ok_or(SpecificError::MissingIndex)?;
+
+        // Determine the appropriate version based on block type
+        let block_version = match &block {
```
how about sth like 

```rust
impl<E: EthSpec, P: AbstractExecPayload<E>> TryFrom<BeaconBlock<E, P>> for DataVersion {
    type Error = UnknownDataVersion;

    fn try_from(value: BeaconBlock<E, P>) -> Result<DataVersion, Self::Error> {
        match value {
            BeaconBlock::Base(_) => Ok(DATA_VERSION_PHASE0),
            BeaconBlock::Altair(_) => Ok(DATA_VERSION_ALTAIR),
            BeaconBlock::Bellatrix(_) => Ok(DATA_VERSION_BELLATRIX),
            BeaconBlock::Capella(_) => Ok(DATA_VERSION_CAPELLA),
            BeaconBlock::Deneb(_) => Ok(DATA_VERSION_DENEB),
            BeaconBlock::Electra(_) => DATA_VERSION_ELECTRA,
            _ => Err(UnknownDataVersion),
        }
    }
}
```

dknopik replied:
I've reworked the `DataVersion` a bit, using `ForkName` internally. Looking forwards to your thoughts about that approach, I am a bit on the fence about it.

diegomrsantos commented on anchor/validator_store/src/lib.rs:
```diff
@@ -459,10 +460,21 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+            Ok(DataSsz::BeaconBlock(block))
+        } else {
+            error!(%fork, "Failed to deserialize decided block");
+            Err(Error::SpecificError(SpecificError::InvalidQbftData))
```
Is trial and error the only way to do that? The previous code also checked for more cases, wasn't it necessary?

dknopik replied:
Yes, trial and error is the only way as the surrounding structure does not indicate whether the block is blinded or not.

dknopik replied:
The previous code tried all `DataSsz` variants, uselessly, as we would fail anyway in the calling function if an invalid variant would be encountered.

diegomrsantos commented on anchor/validator_store/src/lib.rs:
```diff
@@ -1038,11 +1050,16 @@ impl<T: SlotClock, E: EthSpec> ValidatorStore for AnchorValidatorStore<T, E> {
-        let message = match data_ssz {
-            Ok(DataSsz::AggregateAndProof(message)) => message,
-            _ => return Err(Error::SpecificError(SpecificError::InvalidQbftData)),
+        let message = if data.version < DATA_VERSION_ELECTRA {
```
We could use a match here

dknopik replied:
personally, for decisions with 2 options, i prefer a classic if-else.

as noted elsewhere, the data version should be a enum anyway, so i will need to touch this again. as soon as this is an enum, I might need to change it anyway :)

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -98,8 +98,11 @@ where
-    /// The current round this instance state is in.a
+    /// The current round this instance state is in.
     current_round: Round,
+    /// The round requested by the user. Used to not accidentally start a round that was already
```
> The round requested by the user

Could you elaborate on what this means?

Zacholme7 replied:
Second this

dknopik replied:
User, as in user of the crate. The user calls `end_round` if they want the round they requested previously to end, usually timer-based. They previously requested a round via initialization config or another `end_round` call.

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -159,6 +162,7 @@ where
             valid_start_data,
             data: HashMap::new(),
             current_round,
+            requested_round: current_round,
```
Should this be `None` at this point?

dknopik replied:
No.

By initializing the instance, the user sets the instance into the first round as requested in the config, therefore the first round is the requested one

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -832,14 +836,17 @@ where
-
-        self.send_round_change(Hash256::default());
-        self.start_round();
+        self.requested_round = next_round;
```
`requested_round` is initially equal to `current_round` and this seems to be the only place where `requested_round` is updated. How is the value updated by this "request"?

dknopik replied:
It is incremented by one, see `next` above

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -832,14 +836,17 @@ where
-        self.send_round_change(Hash256::default());
-        self.start_round();
+        self.requested_round = next_round;
+        if self.current_round < next_round {
```
Why is the `current_round` updated only conditionally now?

dknopik replied:
There is an explanation of the issue in the PR description - I am not sure which part of it is unclear, please let me know which aspect I should elaborate more.

Zacholme7 reviewed (commented):
Diego covered pretty much everything. Also just seconded some small terminology clarification on my end.

dknopik committed:
better naming and docs

dknopik committed:
fmt

dknopik committed:
Use current slot for RANDAO Reveal

dknopik committed:
Merge branch 'refs/heads/unstable' into proposal-fixes

# Conflicts:
#	anchor/validator_store/src/lib.rs

dknopik committed:
Address review and fix proposal data

dknopik commented:
Hey @diegomrsantos and @Zacholme7,

I found and solved more issues (see description), and made the code around versions a bit cleaner (IMO). Let me know if you have any questions or suggestions!

As one fix requires a (maybe slightly questionable) change in LH, I'd like to wait until that is approved upstream, but feel free to review already. :) Thank you!

dknopik committed:
Merge branch 'unstable' into proposal-fixes

dknopik committed:
Custom LH until upstream PR is merged...

diegomrsantos commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -226,48 +226,81 @@ impl TreeHash for BeaconRole {
+    fn from_ssz_bytes(bytes: &[u8]) -> Result<Self, DecodeError> {
+        let num = u64::from_ssz_bytes(bytes)?;
+        match num {
+            0 => Ok(DataVersion::Unknown),
```
How does the code proceed if we hit this?

diegomrsantos commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -226,48 +226,81 @@ impl TreeHash for BeaconRole {
-            DATA_VERSION_ELECTRA => Ok(ForkName::Electra),
-            _ => Err(()),
+            DataVersion::Fork(fork) => Ok(fork),
+            _ => Err(UnknownDataVersion),
```
If that's the answer to my question above, can we get rid of the unknown variant and return an error like this during decoding?

dknopik replied:
Hmm, a "Unknown" version is still well defined with a value of "0" and is different case from a value that is not well defined at all (such as "17"). This might be a bit pedantic though as it seems to be unused in SSV.

Eh, lets get rid of it.

diegomrsantos replied:
Thank you. Perhaps one last improvement would be adding a comment explaining why we can't use `ForkName` directly. If I saw the code for the first time, that would be the question in my mind.

diegomrsantos replied:
@dknopik has this been done?

dknopik replied:
Sorry - forgot this. Thanks for pinging me. I've added a comment, please take a look if it is sufficient

diegomrsantos replied:
It's great, thank you.

dknopik commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -227,33 +226,82 @@ impl TreeHash for BeaconRole {
-#[ssz(struct_behaviour = "transparent")]
-pub struct DataVersion(u64);
+#[derive(Clone, Debug, PartialEq, Eq, PartialOrd, Ord)]
+pub enum DataVersion {
```
It would not compile if we truly needed it m8

dknopik committed:
Further simplify `DataVersion`

dknopik committed:
Remove `DataSsz`

It did not really provide value anymore since we started decoding the correct objects directly.

dknopik committed:
Move max transmit size to a `const`

dknopik committed:
Merge branch 'unstable' into proposal-fixes

Zacholme7 reviewed (commented):
lgtm, ill give @jking-aus final review

dknopik committed:
Merge branch 'unstable' into proposal-fixes

dknopik committed:
Improve fix for round timeouts

dknopik commented:
pls do not merge yet, I want to properly update the LH deps in a separate PR

diegomrsantos commented on anchor/qbft_manager/src/instance.rs:
```diff
@@ -177,6 +171,12 @@ impl<D: QbftData> From<Option<QbftMessage<D>>> for RecvResult<D> {
     async fn recv(&mut self, rx: &mut UnboundedReceiver<QbftMessage<D>>) -> RecvResult<D> {
+        // We calculate the sleep dynamically, as both messages and the local timer might cause the
+        // round to advance
+        let round_end = calculate_round_timeout(self.qbft.get_round().into(), &self.start_time);
```
Does it mean `round_end_time` or `round_timeout`?

dknopik replied:
seeing that the type is `Instant`, there is not really a difference IMO. But always happy to considers suggestions for better naming

diegomrsantos commented on anchor/qbft_manager/src/instance.rs:
```diff
@@ -185,11 +185,7 @@ impl<D: QbftData<Hash = Hash256>> Initialized<D> {
-                let next_round = self.qbft.get_round().get() as u64 + 1_u64;
-                let sleep = calculate_round_timeout(next_round, &self.start_time);
-                self.round_end = Box::pin(tokio::time::sleep_until(sleep));
+            _ = &mut round_timeout_sleep => {
```
Is there any behaviour change in extracting the logic, or is it only to improve readability? Why was the `+ 1` removed?

dknopik replied:
yes, there is a change: by extracting the logic we make sure to catch any round changes caused by messages and time out at the end of the new round if so

The `+ 1` is removed because we are now calculating the timeout for the current round, and not for the "next" round as done by the previous code

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -842,17 +837,14 @@ where
-            self.start_round();
-        }
+        self.send_round_change(Hash256::default());
+        self.start_round();
```
I wouldn't expect this to be called inside this function.

dknopik replied:
this function is not modified anymore by this PR in total (only by the latest commit), so it is not really in scope for this review

diegomrsantos commented:
@dknopik, can we add tests to be sure (and avoid regressions) that there's no race related to the round timeout?

dknopik commented on anchor/validator_store/src/lib.rs:
```diff
@@ -382,72 +382,60 @@ impl<T: SlotClock, E: EthSpec> AnchorValidatorStore<T, E> {
+    ) -> Result<UnsignedBlock<E>, Error> {
         let validator = self.validator(validator_pubkey)?;
 
+        let block = signable_block.as_block();
```
Nah, this sanity check is only needed before signing

dknopik commented on anchor/validator_store/src/lib.rs:
```diff
@@ -1006,8 +1016,8 @@ impl<T: SlotClock, E: EthSpec> ValidatorStore for AnchorValidatorStore<T, E> {
-            Attestation::Base(_) => DATA_VERSION_PHASE0,
-            Attestation::Electra(_) => DATA_VERSION_ELECTRA,
+            Attestation::Base(_) => ForkName::Base.into(),
+            Attestation::Electra(_) => ForkName::Electra.into(),
```
There are only these two types of attestations

dknopik commented:
> @dknopik, can we add tests to be sure (and avoid regressions) that there's no race related to the round timeout?

Sure. Do you want me to do so? If you want, you can take a stab at it instead - that might prevent issues like me "overfitting" the test on my mental model of the issue.

dknopik committed:
Merge branch 'unstable' into proposal-fixes

dknopik committed:
Merge branch 'unstable' into proposal-fixes

diegomrsantos commented:
> > @dknopik, can we add tests to be sure (and avoid regressions) that there's no race related to the round timeout?
> 
> Sure. Do you want me to do so? If you want, you can take a stab at it instead - that might prevent issues like me "overfitting" the test on my mental model of the issue.

Feel free to do it.

diegomrsantos reviewed (dismissed):
Thanks for the work on this one, it's really great.

dknopik committed:
Add comment explaining `DataVersion`

dknopik committed:
reintroduce round check in qbft

dknopik committed:
add test for data race issue

dknopik commented:
@diegomrsantos thanks for the idea about a test - I found another minor data race thanks to it.

We removed the message round check in the qbft crate as we now check for this in message validation. However, as the message might be handled by the qbft crate after the round has advanced, this caused an interesting issue where a message belonging to a previous round change arrived and caused an instance to revert to a previous round.

I reintroduced the round check.

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -235,6 +235,19 @@ where
+        if (wrapped_msg.qbft_message.round < current_round as u64)
+            || (wrapped_msg.qbft_message.round > self.config.max_rounds() as u64)
+        {
+            warn!(
```
Do you think it should be a `warn`?

dknopik replied:
nope, fixed

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -235,6 +235,19 @@ where
+            warn!(
+                propose_round = wrapped_msg.qbft_message.round,
+                current_round = *self.current_round,
+                "Message received for a invalid round"
```
How about "for a previous round"?

dknopik replied:
changed message and removed check for `max_round`, as we check that in validation and is not subject to a data race

diegomrsantos commented on anchor/qbft_manager/src/tests.rs:
```diff
@@ -110,21 +112,37 @@ where
+                    );
+                },
+                _ = &mut timeout => {
+                    panic!("test timed out")
```
How about defining the timeout here?

dknopik replied:
There is a difference between defining it here vs above. If we define it here, it is refreshed every time a message comes in from the queue. If we define it above, it is only created once, so we wait at most 5 minutes. Personally, I prefer the latter as it is more predictable, but I don't feel strongly.

diegomrsantos replied:
Or even better, a const in a more visible place.

dknopik committed:
address review

diegomrsantos commented on anchor/qbft_manager/src/tests.rs:
```diff
@@ -653,6 +680,7 @@ mod manager_tests {
             setup.executor,
             CommitteeSize::Four,
             setup.all_data,
+            HashMap::new(),
```
I've been thinking about this. Unfortunately, Rust doesn't have default parameters. Could you tell me what's the idiomatic way to handle this? Creating an additional function `new_with_delay`?

dknopik replied:
if there is only only one "optional" parameter, `new_with_x` is fine. However, once you have more optional parameters, a builder struct is probably the way to go, because `new_with_x_and_y` gets ugly fast unless you have a nice way of phrasing the different `new_*` functions in the context of the application. I'll add `new_with_delay`.

diegomrsantos commented on anchor/qbft_manager/src/tests.rs:
```diff
@@ -828,6 +865,31 @@ mod manager_tests {
+            setup.all_data,
+            HashMap::from([
+                (OperatorId(2), Duration::from_secs(3)),
+                (OperatorId(3), Duration::from_secs(5)),
```
Very nice docs at the beginning. How about extracting it to a new variable with a descriptive name? It might be useful to state the round time too.

diegomrsantos replied:
Looks good, thanks

dknopik committed:
more address review

dknopik commented on anchor/common/ssv_types/src/consensus.rs:
```diff
@@ -227,75 +227,61 @@ impl TreeHash for BeaconRole {
-    fn tree_hash_packed_encoding(&self) -> PackedEncoding {
-        self.0.tree_hash_packed_encoding()
+    fn ssz_append(&self, buf: &mut Vec<u8>) {
+        let num: u64 = match self.0 {
+            ForkName::Base => 1,
+            ForkName::Altair => 2,
+            ForkName::Bellatrix => 3,
+            ForkName::Capella => 4,
+            ForkName::Deneb => 5,
+            ForkName::Electra => 6,
+            ForkName::Fulu => 7,
+        };
```
no

dknopik committed:
Merge branch 'unstable' into proposal-fixes

diegomrsantos commented on anchor/common/qbft/src/lib.rs:
```diff
@@ -235,6 +235,17 @@ where
     ) -> Option<(Option<ValidData<D>>, OperatorId)> {
+        // Ensure that this message is for the correct round
+        let current_round = self.current_round.get();
+        if wrapped_msg.qbft_message.round < current_round as u64 {
```
I think we can use into() instead

dknopik replied:
No, there is no `From<usize> for u64`

This kind of discrepancy between our message types and our custom types such as `Round` is something I have been meaning to fix for a while... When that PR lands, this will be nicer ;)

dknopik committed:
nicer conversion
